website/api/controllers/account/update-password.js

```diff
@@ -42,6 +46,37 @@ module.exports = {
     // Hash the new password.
     var hashed = await sails.helpers.passwords.hashPassword(inputs.newPassword);
 
+    // If this user has a valid provisioned Fleet Sandbox instance, we'll update their password there as well.
+    if(this.req.me.fleetSandboxURL) {
+
+      // If the user's Fleet Sandbox instance is still valid, we'll use their old password to get the authorization token from the sandbox and use it to change their password.
+      if(this.req.me.fleetSandboxExpiresAt < Date.now()) {
+        // Get the record for this user to update their Fleet Sandbox password using their old hashed password.
+        let userRecord = await User.findOne({ id: this.req.me.id });
+
+        // Send a post request to the `/login` endpoint of the Fleet Sandbox instance to get this user's api token.
+        let authToken = await sails.helpers.http.post(userRecord.fleetSandboxURL+'/api/v1/fleet/login', {
+          'email': userRecord.emailAddress,
+          'password': userRecord.password
+        }).intercept('non200Response', 'couldNotChangeSandboxPassword');
+
+        // If we received a token in the response from the Fleet Sandbox instance, we'll use that to send a POST request to the `/change_password` endpoint to update this users password with the hashed version of their fleetdm.com password
+        if(!authToken.token) {
+          throw 'couldNotChangeSandboxPassword';
+        } else {
+          // Update the user's password on their fleet instance
+          await sails.helpers.http.post(
+              userRecord.fleetSandboxURL+'/api/v1/fleet/change_password',
+            {
+              'old_password': userRecord.password,
+              'new_password': hashed,
+            },
+            {'Authorization': 'Bearer '+authToken.token}
+          ).intercept('non200Response', 'couldNotChangeSandboxPassword');
+        }
+      }
+    }
+
```

Comment:
My understanding is that their fleetdm.com and password is the only way users will ever access Fleet Sandbox, so changes to that password don't need to be sent to the Fleet instance.

If there was a change to that requirement or you think there could be confusion, could you please let @noahtalerman know ASAP?

Reply:
@mikermcneil The Fleet Sandbox instance needs the user's email address and (hashed) password to log the user in with the `/demologin` endpoint. 
If the user changes their password on fleetdm.com, they wouldn't be able to log in to their Fleet Sandbox instance with the credentials that are sent to the `/demologin` endpoint.

Reply:
Ah, I see.  So this is another side effect of the change to the authentication strategy that was made earlier this month.  The user would never actually know they are using those credentials to sign in, it’s just that fleetdm.com is using those credentials behind the scenes instead of a demoKey.

@zhumo FYI this means it’ll be important to hide the “Change my password” functionality in the Fleet product, when it is running as part of Fleet Sandbox.  I don’t remember seeing it in the wireframes, because it didn’t matter previously before the underlying authentication mechanism was changed a week or so ago.

Reply:
@mikermcneil I realized when I was typing that there was a different/possible better way to do this.

How does this sound:

- When signing up, a `fleetSandboxDemoKey` string is created, using a combination of `sails.helpers.strings.random('url-friendly', 12)`, The new user's email address, and the sandbox expiration JS timestamp. It uses a combination of those three values so that we can make sure that every `fleetSandboxDemoKey` we send to the Cloud Provisioner API meets the Fleet password requirements
- The `fleetSandboxDemoKey` is used in place of the User's (hashed) password when provisioning a Fleet Sandbox Instance
- This value is stored on the User model and is only used to log the user into their Fleet Sandbox instance, which will only be valid for 24 hours.

If I make that change, we can remove this portion of the code **and** we won't need update our password requirements on fleetdm.com (which we probably don't need to do regardless, because we using hashed versions of the users password to create their Fleet Sandbox user)

Reply:
My understanding of Eric's proposal is that we would have the user log in to Fleetdm.com, which will retrieve this `fleetSandboxDemoKey`, and secretly in the back end, we are using that key to authenticate them into fleet sandbox. This way, they do not need to maintain two separate credentials? 

If that's the correct understanding, then that sounds good to me from a product perspective.

Reply:
@eashaw I like your thinking, in terms of not sending a real hashed password to avoid confusion.  What would happen if the email is changed on fleetdm.com?  Does email need to be included in the demo key or can it be excluded and some other special character and number included instead?  Is there any reason to include the expiry in this other than to make sure there’s a number?

@zhumo either way, disabling changing one’s own email and password in the Fleet product for sandbox is probably still a good idea, because otherwise users might misunderstand. Also if they are able to change the password, they could still break their ability to access the sandbox.

Reply:
>Does email need to be included in the demo key or can it be excluded and some other special character and number included instead? Is there any reason to include the expiry in this other than to make sure there’s a number?

Those would only for the guaranteed number and special character. We could also just add an arbitrary number and special character to the random string.

Reply:
@mikermcneil I opted to use `sails.helpers.strings.uuid()` to generate the `fleetSandboxDemoKey` because it meets all the password requirements of the Sandbox Instance without being changed.

Reply:
@mikermcneil I think it's already been hidden per: https://github.com/fleetdm/fleet/issues/6544.

The other place where this might happen is if you do forgot pw. If we are using fleetdm.com's credentials in place of fleet sandbox's credentials, then all forgot pw stuff should be handled via fleetdm.com not sandbox. 

@eashaw , I see that you are assigned to this one: https://github.com/fleetdm/fleet/issues/5036. Will this handle the forgot password flow in a way such that a user can't reset their sandbox pw?

Reply:
@zhumo Yes! The forgot password flow will only handle fleetdm.com credentials. The password for the Fleet Sandbox instance would be stored as a separate attribute on the User model and can't be reset.